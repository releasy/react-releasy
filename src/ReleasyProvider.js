// @flow
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { Environment } from 'relay-runtime';
import invariant from 'invariant';

import { getEnvironment } from './store';

type Props = {
  children: React.Node,
};

class ReleasyProvider extends React.PureComponent<Props> {
  static propTypes = {
    children: PropTypes.element.isRequired,
  };

  static childContextTypes = {
    environment: PropTypes.instanceOf(Environment).isRequired,
  };

  constructor(props: Props) {
    invariant(props.children, 'ReleasyProvider: missing "children"');

    super(props);
  }

  getChildContext() {
    return {
      environment: getEnvironment(),
    };
  }

  render() {
    return this.props.children;
  }
}

export default ReleasyProvider;
