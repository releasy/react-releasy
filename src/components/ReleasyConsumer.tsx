import * as React from 'react';

import GlobalContext, { GlobalContextValueType } from '../contexts/GlobalContext';

type ReleasyConsumerPropsType = {
  children: React.ReactNode,
};

const ReleasyConsumer = ({
  children,
}: ReleasyConsumerPropsType): React.ReactElement<GlobalContextValueType> => (
  <GlobalContext.Consumer>
    {(globalContextValue: GlobalContextValueType) =>
      React.Children.map(children, (child: React.ReactElement<any>) =>
        React.cloneElement(child, globalContextValue),
      )
    }
  </GlobalContext.Consumer>
);

export default ReleasyConsumer;
