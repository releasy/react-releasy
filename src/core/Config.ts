import { CacheConfig, RequestNode, UploadableMap, Variables, GraphQLResponse } from 'relay-runtime';
import invariant from 'invariant';

export interface LinkInterface {
  fetch: (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables: UploadableMap,
  ) => Promise<GraphQLResponse>;
}

export interface CacheInterface {
  get: (queryID: string, variables: Variables) => GraphQLResponse;
  set: (queryID: string, variables: Variables, payload: GraphQLResponse) => void;
  clear: () => void;
  size: () => number;
}

type ConfigPropsType = {
  link: LinkInterface,
  cache?: CacheInterface,
  devTools?: boolean,
  networkLogger?: boolean,
};

class Config {
  link: LinkInterface;
  cache: CacheInterface;
  devTools: boolean;
  networkLogger: boolean;

  constructor(props: ConfigPropsType) {
    invariant(props, 'Config: missing props');
    invariant(props.link, 'Config: missing "link"');

    this.link = props.link;
    this.cache = props.cache || null;
    this.devTools = props.devTools || false;
    this.networkLogger = props.networkLogger || false;
  }
}

export default Config;
