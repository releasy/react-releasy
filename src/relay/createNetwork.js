// @flow
import { Network } from 'relay-runtime';
import RelayNetworkLogger from 'relay-runtime/lib/RelayNetworkLogger';

import type { CacheConfig, RequestNode, UploadableMap, Variables } from 'relay-runtime';

import { isMutation, mustForceFetch } from './utils';

import type { Config } from '../index';

const createNetwork = (config: Config): Network => {
  const fetchFn = config.networkLogger
    ? RelayNetworkLogger.wrapFetch(config.link.fetch)
    : config.link.fetch;

  return Network.create(
    async (
      request: RequestNode,
      variables: Variables,
      cacheConfig: CacheConfig,
      uploadables?: UploadableMap,
    ) => {
      const queryID = request.text;

      // no cache configured, just fetch the data from server
      if (!config.cache) {
        return fetchFn(request, variables, cacheConfig, uploadables);
      }

      // on mutations, clear cache and then fetch the data
      if (isMutation(request)) {
        config.cache.clear();
        return fetchFn(request, variables, cacheConfig, uploadables);
      }

      // verify if data exists on cache
      const fromCache = config.cache.get(queryID, variables);
      if (fromCache !== null && !mustForceFetch(cacheConfig)) {
        return fromCache;
      }

      const fromServer = await fetchFn(request, variables, cacheConfig, uploadables);
      if (fromServer) {
        config.cache.set(queryID, variables, fromServer);
      }

      return fromServer;
    }
  );
};

export default createNetwork;
