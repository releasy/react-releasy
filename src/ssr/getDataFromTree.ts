export enum ComponentType {
  CONTEXT_PROVIDER = 'React.Provider',
  CONTEXT_CONSUMER = 'React.Consumer',
  CLASS_COMPONENT = 'React.ComponentClass',
  STATELESS_COMPONENT = 'React.SFC',
  DOM_ELEMENT = 'React.DOMElement',
}

const isContextProvider = (element: React.ReactNode): element is React.Provider<any> => {
  const type = (element as any).type;
  return !!type && !!type._context;
};

const isContextConsumer = (element: React.ReactNode): element is React.Consumer<any> => {
  const type = (element as any).type;
  return !!type && !!type.Consumer;
};

const isClassComponent = (element: React.ReactNode): element is React.ComponentClass<any> => {
  const type = (element as any).type;
  return !!type && type.prototype &&
    (type.prototype.isReactComponent ||
      type.prototype.isPureReactComponent);
};

const isStatelessComponent = (element: React.ReactNode): element is React.SFC<any> => {
  const type = (element as any).type;
  return !!type && typeof type === 'function';
};

const isDOMElement = (element: React.ReactNode): element is React.DOMElement<any, any> => {
  return !isContextProvider(element) &&
    !isContextConsumer(element) &&
    !isClassComponent(element) &&
    !isStatelessComponent(element);
};

export const identifyComponentType = (element: React.ReactNode): ComponentType => {
  const rules = {
    [ComponentType.CONTEXT_PROVIDER]: isContextProvider,
    [ComponentType.CONTEXT_CONSUMER]: isContextConsumer,
    [ComponentType.CLASS_COMPONENT]: isClassComponent,
    [ComponentType.STATELESS_COMPONENT]: isStatelessComponent,
    [ComponentType.DOM_ELEMENT]: isDOMElement,
  };

  return Object.keys(ComponentType).reduce(
    (accumulator, key) => {
      const value = ComponentType[key];
      const rule = rules[value];

      if (accumulator || !rule(element)) {
        return accumulator;
      }

      return value;
    },
    null,
  );
};

const handleContextProvider = async (
  element: React.Provider<any>,
  context: any,
  visitor: (element: React.ReactNode, instance: React.Component<any>) => void,
) => {
  const type = (element as any).type;
  const props = (element as any).props;

  const currentValue = props.value || {};
  type._context.currentValue = currentValue;

  await visitor(element, null);

  return walkTree(
    props.children,
    context,
    visitor,
  );
};

const handleContextConsumer = async (
  element: React.Consumer<any>,
  context: any,
  visitor: (element: React.ReactNode, instance: React.Component<any>) => void,
) => {
  const type = (element as any).type;
  const props = (element as any).props;

  await visitor(element, null);

  return walkTree(
    props.children(type.currentValue),
    context,
    visitor,
  );
};

const handleClassComponent = async (
  element: React.ComponentClass<any>,
  context: any,
  visitor: (element: React.ReactNode, instance: React.Component<any>) => void,
) => {
  const type = (element as any).type;

  const props = {
    ...type.defaultProps,
    ...(element as any).props,
  };

  const instance = new type(props, context);

  instance.context = instance.context || context;
  instance.state = instance.state || null;

  instance.setState = (state) => {
    const newState = typeof state === 'function' ? state(instance.state, instance.props) : state;
    instance.state = { ...instance.state, ...newState };
  };

  if (type.getDerivedStateFromProps) {
    const newState = type.getDerivedStateFromProps(instance.props, instance.state);
    instance.state = { ...instance.state, ...newState };
  } else if (instance.UNSAFE_componentWillMount) {
    instance.UNSAFE_componentWillMount();
  } else if (instance.componentWillMount) {
    instance.componentWillMount();
  }

  const newContext =  instance.getChildContext ? {
    ...context,
    ...instance.getChildContext(),
  } : context;

  await visitor(element, instance);

  return walkTree(
    instance.render(),
    newContext,
    visitor,
  );
};

const handleStatelessComponent = async (
  element: React.SFC<any>,
  context: any,
  visitor: (element: React.ReactNode, instance: React.Component<any>) => void,
) => {
  const type = (element as any).type;

  const props = {
    ...type.defaultProps,
    ...(element as any).props,
  };

  await visitor(element, null);

  return walkTree(
    type(props),
    context,
    visitor,
  );
};

const handleDOMElement = async (
  element: React.DOMElement<any, any>,
  context: any,
  visitor: (element: React.ReactNode, instance: React.Component<any>) => void,
) => {
  const props = (element as any).props || {};

  await visitor(element, null);

  return walkTree(
    props.children,
    context,
    visitor,
  );
};

export const walkTree = (
  element: React.ReactNode,
  context: any = {},
  visitor: (element: React.ReactNode, instance: React.Component<any>) => void,
) => {
  if (!element) {
    return;
  }

  if (Array.isArray(element)) {
    return Promise.all(element.map(item => walkTree(item, context, visitor)));
  }

  const handlers = {
    [ComponentType.CONTEXT_PROVIDER]: handleContextProvider,
    [ComponentType.CONTEXT_CONSUMER]: handleContextConsumer,
    [ComponentType.CLASS_COMPONENT]: handleClassComponent,
    [ComponentType.STATELESS_COMPONENT]: handleStatelessComponent,
    [ComponentType.DOM_ELEMENT]: handleDOMElement,
  };

  const componentType = identifyComponentType(element);
  const handler: any = handlers[componentType];

  return handler(element, context, visitor);
};

// @TODO - improve this later
const resolveRelayQueryRenderer = (instance: React.Component<any>) => {
  const { environment, query, variables } = instance.props;
  const { getRequest } = environment.unstable_internal;
  const request = getRequest(query);

  return new Promise((resolve) => {
    const interval = setInterval(
      () => {
        const { _snapshot } = (instance.state as any).queryFetcher;
        if (!_snapshot) {
          return;
        }

        const { data } = _snapshot;

        clearInterval(interval);
        resolve({ request, variables, data });
      },
      0,
    );
  });
};

const getDataFromTree = async (
  rootElement: React.ReactNode,
  rootContext: any = {},
): Promise<any> => {
  const promises = [];

  await walkTree(
    rootElement,
    rootContext,
    async (element: React.ReactNode, instance: React.Component<any>) => {
      const { type } = (element as any);

      if (!type || type.name !== 'ReactRelayQueryRenderer') {
        return;
      }

      const promise = resolveRelayQueryRenderer(instance);
      promises.push(promise);
    },
  );

  return Promise.all(promises);
};

export default getDataFromTree;
