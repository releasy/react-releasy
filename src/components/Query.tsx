import * as React from 'react';
import { QueryRenderer } from 'react-relay';
import { GraphQLTaggedNode, Variables } from 'relay-runtime';

import { DataFrom } from '../types/react-relay.d';
import GlobalContext, { GlobalContextValueType } from '../contexts/GlobalContext';
import Config from '../core/Config';

export type QueryChildrenPropsType = {
  error?: Error,
  retry: Function,
  isFetching: boolean,
};

type QueryPropsType = {
  query: GraphQLTaggedNode,
  variables?: Variables | Function,
  children: (props: QueryChildrenPropsType) => JSX.Element,
  dataFrom?: DataFrom,
};

const Query = (props: QueryPropsType): React.ReactElement<any> => {
  const {
    query,
    variables,
    children,
    dataFrom,
  } = props;

  const getVariables = (): Variables => {
    if (typeof variables === 'function') {
      return variables(props);
    }

    return variables;
  };

  const getDataFrom = (config: Config): DataFrom => {
    if (!dataFrom && config.ssrMode === true) {
      return DataFrom.STORE_THEN_NETWORK;
    }

    return dataFrom;
  };

  return (
    <GlobalContext.Consumer>
      {({ environment, config }: GlobalContextValueType) =>
        <QueryRenderer
          environment={environment}
          query={query}
          variables={getVariables()}
          dataFrom={getDataFrom(config)}
          render={({ props: queryProps, error, retry }) => {
            return children({
              ...queryProps,
              error,
              retry,
              isFetching: !queryProps && !error,
            });
          }}
        />
      }
    </GlobalContext.Consumer>
  );
};

export default Query;
