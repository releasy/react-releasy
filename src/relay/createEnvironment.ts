import { installRelayDevTools } from 'relay-devtools';
import { Environment, RecordSource, Store } from 'relay-runtime';

import Config from '../core/Config';
import createNetwork from './createNetwork';

const createEnvironment = (config: Config): Environment => {
  if (config.devTools) {
    installRelayDevTools();
  }

  const network = createNetwork(config);

  const source = new RecordSource();
  const store = new Store(source);

  return new Environment({
    network,
    store,
  });
};

export default createEnvironment;
