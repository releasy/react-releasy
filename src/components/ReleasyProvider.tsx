import * as React from 'react';
import invariant from 'invariant';

import GlobalContext, {
  GlobalContextValueType,
  setConfig,
  setEnvironment,
} from '../contexts/GlobalContext';
import Config from '../core/Config';
import createEnvironment from '../relay/createEnvironment';

type ReleasyProviderPropsType = {
  config: Config,
  children: React.ReactNode,
};

type ReleasyProviderStateType = {
  globalContextValue: GlobalContextValueType,
};

class ReleasyProvider extends React.PureComponent<
  ReleasyProviderPropsType,
  ReleasyProviderStateType
> {
  constructor(props: ReleasyProviderPropsType) {
    super(props);

    invariant(props.config, 'ReleasyProvider: missing "config"');

    const { config } = props;
    const environment = createEnvironment(config);

    setConfig(config);
    setEnvironment(environment);

    this.state = {
      globalContextValue: {
        config,
        environment,
      },
    };
  }

  render() {
    const { children } = this.props;
    const { globalContextValue } = this.state;

    return (
      <GlobalContext.Provider value={globalContextValue}>
        {children}
      </GlobalContext.Provider>
    );
  }
}

export default ReleasyProvider;
