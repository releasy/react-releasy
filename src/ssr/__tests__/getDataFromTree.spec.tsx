import * as React from 'react';

import { MeQuery } from '../../../tests/components';
import { createConfig } from '../../../tests/fixtures/core';
import { createQueryResponse } from '../../../tests/fixtures/relay';

import ReleasyProvider from '../../components/ReleasyProvider';
import getDataFromTree, {
  ComponentType,
  identifyComponentType,
  walkTree,
} from '../getDataFromTree';

it('should identify context', () => {
  const initialValue = { name: 'John' };

  const MyContext = React.createContext(initialValue);

  expect(identifyComponentType(
    <MyContext.Provider value={initialValue}>
      <div />
    </MyContext.Provider>,
  )).toBe(ComponentType.CONTEXT_PROVIDER);

  expect(identifyComponentType(
    <MyContext.Consumer>
      {() => <div />}
    </MyContext.Consumer>,
  )).toBe(ComponentType.CONTEXT_CONSUMER);
});

it('should identify class component', () => {
  class MyComponent extends React.Component {
    render() {
      return this.props.children;
    }
  }

  expect(identifyComponentType(
    <MyComponent>
      <div />
    </MyComponent>,
  )).toBe(ComponentType.CLASS_COMPONENT);
});

it('should identify stateless component', () => {
  const MyComponent = props => props.children;

  expect(identifyComponentType(
    <MyComponent>
      <div />
    </MyComponent>,
  )).toBe(ComponentType.STATELESS_COMPONENT);
});

it('should identify dom elements', () => {
  expect(identifyComponentType(
    <div />,
  )).toBe(ComponentType.DOM_ELEMENT);

  expect(identifyComponentType(
    <span />,
  )).toBe(ComponentType.DOM_ELEMENT);

  expect(identifyComponentType(
    <p />,
  )).toBe(ComponentType.DOM_ELEMENT);
});

it('should walk context', async () => {
  const initialValue = { name: 'John' };
  const realValue = { name: 'Doe' };

  const MyContext = React.createContext(initialValue);

  const components = [];

  await walkTree(
    <MyContext.Provider value={realValue}>
      <MyContext.Consumer>
        {(anotherValue: object) => {
          expect(anotherValue).toBe(realValue);

          return [<div />, <div />];
        }}
      </MyContext.Consumer>
    </MyContext.Provider>,
    {},
    (element: React.ReactNode) => {
      const componentType = identifyComponentType(element);
      components.push(componentType);
    },
  );

  expect(components).toMatchSnapshot();
});

it('should walk class component', async () => {
  class MyComponent extends React.Component {
    render() {
      return (
        <div />
      );
    }
  }

  const components = [];

  await walkTree(
    <MyComponent />,
    {},
    (element: React.ReactNode) => {
      const componentType = identifyComponentType(element);
      components.push(componentType);
    },
  );

  expect(components).toMatchSnapshot();
});

it('should walk class component with getDerivedStateFromProps', async () => {
  class MyComponent extends React.Component {
    static getDerivedStateFromProps() {
      return { name: 'John' };
    }

    render() {
      return (
        <div />
      );
    }
  }

  const components = [];

  await walkTree(
    <MyComponent />,
    {},
    (element: React.ReactNode) => {
      const componentType = identifyComponentType(element);
      components.push(componentType);
    },
  );

  expect(components).toMatchSnapshot();
});

it('should walk class component with UNSAFE_componentWillMount', async () => {
  class MyComponent extends React.Component {
    UNSAFE_componentWillMount() {
      this.setState({ name: 'John' });
    }

    render() {
      return (
        <div />
      );
    }
  }

  const components = [];

  await walkTree(
    <MyComponent />,
    {},
    (element: React.ReactNode) => {
      const componentType = identifyComponentType(element);
      components.push(componentType);
    },
  );

  expect(components).toMatchSnapshot();
});

it('should walk class component with componentWillMount', async () => {
  class MyComponent extends React.Component {
    componentWillMount() {
      this.setState(() => ({ name: 'John' }));
    }

    render() {
      return (
        <div />
      );
    }
  }

  const components = [];

  await walkTree(
    <MyComponent />,
    {},
    (element: React.ReactNode) => {
      const componentType = identifyComponentType(element);
      components.push(componentType);
    },
  );

  expect(components).toMatchSnapshot();
});

it('should walk stateless component', async () => {
  const MyComponent = () => <div />;

  const components = [];

  await walkTree(
    <MyComponent />,
    {},
    (element: React.ReactNode) => {
      const componentType = identifyComponentType(element);
      components.push(componentType);
    },
  );

  expect(components).toMatchSnapshot();
});

it('should get data from tree', async () => {
  global.fetch.mockResponse(
    JSON.stringify(createQueryResponse()),
    {
      status: 200,
      headers: {
        'content-type': 'application/json',
      },
    },
  );

  const config = createConfig();

  const result = await getDataFromTree(
    <ReleasyProvider config={config}>
      <MeQuery />
    </ReleasyProvider>,
  );

  expect(result).toMatchSnapshot();
});
