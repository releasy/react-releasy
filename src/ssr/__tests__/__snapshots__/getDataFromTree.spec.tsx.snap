// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get data from tree 1`] = `
Array [
  Object {
    "data": Object {
      "me": Object {
        "id": "fakeViewerId",
        "name": "John Doe",
      },
    },
    "dataID": "client:root",
    "node": Object {
      "argumentDefinitions": Array [],
      "kind": "Fragment",
      "metadata": null,
      "name": "MeQuery",
      "selections": Array [
        Object {
          "alias": null,
          "args": null,
          "concreteType": "User",
          "kind": "LinkedField",
          "name": "me",
          "plural": false,
          "selections": Array [
            Object {
              "alias": null,
              "args": null,
              "kind": "ScalarField",
              "name": "id",
              "storageKey": null,
            },
            Object {
              "alias": null,
              "args": null,
              "kind": "ScalarField",
              "name": "name",
              "storageKey": null,
            },
          ],
          "storageKey": null,
        },
      ],
      "type": "Query",
    },
    "seenRecords": Object {
      "client:root": Object {
        "__id": "client:root",
        "__typename": "__Root",
        "me": Object {
          "__ref": "fakeViewerId",
        },
      },
      "fakeViewerId": Object {
        "__id": "fakeViewerId",
        "__typename": "User",
        "id": "fakeViewerId",
        "name": "John Doe",
      },
    },
    "variables": Object {},
  },
]
`;

exports[`should walk class component 1`] = `
Array [
  "React.ComponentClass",
  "React.DOMElement",
]
`;

exports[`should walk class component with UNSAFE_componentWillMount 1`] = `
Array [
  "React.ComponentClass",
  "React.DOMElement",
]
`;

exports[`should walk class component with componentWillMount 1`] = `
Array [
  "React.ComponentClass",
  "React.DOMElement",
]
`;

exports[`should walk class component with getDerivedStateFromProps 1`] = `
Array [
  "React.ComponentClass",
  "React.DOMElement",
]
`;

exports[`should walk context 1`] = `
Array [
  "React.Provider",
  "React.Consumer",
  "React.DOMElement",
  "React.DOMElement",
]
`;

exports[`should walk stateless component 1`] = `
Array [
  "React.SFC",
  "React.DOMElement",
]
`;
